### ISSUE TITLE
Prevent credential steal

### ISSUE THREAD

**Issue by M5oul** [duniter/duniter#233]
Salt and password are stored in clear in `~/.config/ucoin/ucoin_default/ucoin.conf`. If the server is compromised or if someone else got access to the server, credential could give access to someone else to the money we own.

What could uCoin do to prevent that? Encrypt credential? How?

**Comment by c-geek** [comment]
My opinion is that the problem is similar to, if not exactly the same as, SSH keys.

So it's one of two things:

* either the server is considered a trust place and we allow unencrypted storage
* either the server is *not* considered a trust place and we need to manually launch uCoin node which will ask for credentials at runtime.

Note that we could propose the two options during `init`.

**Comment by Insoleet** [comment]
Could we use an interface to password wallets ?
Like Git is doing on linux to store credentials ?

**Comment by c-geek** [comment]
@Insoleet so [git-credential-store](http://git-scm.com/docs/git-credential-store) is just about file permissions apparently. Yes we could use `ucoind store-credentials` interface for example.

It seems Git also uses something similair to GPG agent, they call it [git-credential-cache](http://git-scm.com/docs/git-credential-cache).

If we were able to mimic these agents, I could imagine uCoin to accept being launched *without credentials*, I mean `ucoin start` would work without crashing, but then waiting for such credentials to really being UP and working.

What do you think about it, @Insoleet, @M5oul?

**Comment by Insoleet** [comment]
We also need an "ASK_PASS" feature to be able to interface with Kde Wallet, Gnome Wallet.. :  https://www.kernel.org/pub/software/scm/git/docs/gitcredentials.html

**Comment by Insoleet** [comment]
Maybe we can simply implement an interface to git credentials helpers and use them as password stores : https://www.kernel.org/pub/software/scm/git/docs/technical/api-credentials.html

**Comment by Insoleet** [comment]
Probably the easiest way : 
https://www.npmjs.com/package/git-credential-node
https://www.npmjs.com/package/git-credential-helper

**Comment by c-geek** [comment]
So why not providing our own `ucoin-credentials` algorithm that would notably use UCOIN_ASKPASS, plugged on KDE Wallet / Gnome Wallet for example?

Is it hard to plug on these Wallet tools?

**Comment by Insoleet** [comment]
Oh, this is another good idea too. It depends on the amount of work you wish to do !

**Comment by c-geek** [comment]
*I* wish? Ha! Maybe some strong warriors wanna do it!

**Comment by Insoleet** [comment]
:+1:

**Comment by Sylvain303** [comment]
What about not using password at all, and only signing?
"owner of the node password" must not be stored on the node.

A digital signed certificate can do the trick.

Node's key = NK,
User's private Key, UK, only on his own PC and crypted with a passphrase.

User is signing NK with UK. So NK can now be seen as belonging to User by checking with User's pubUK.

No password stored.

Or I missed something in the topic… I read it quickly.

**Comment by Insoleet** [comment]
We have to sign blocks with the nodes, so we need the private key ! :)

**Comment by Sylvain303** [comment]
I didn't say not to have NK private key. I said to have a **delegated** private key through NK signed by UK. And absolutely not having UK on the node itself, which may compromise UK himself. 

The node as a server can be compromised easily. Don't mix the node with the user running it. 

So for now, do you mean that when you use user’s credential on the node, it also means the node is behaving as if it was the user himself?

So the node is signing directly with UK?
This sounds terribly wrong to me, but I'm too young on the project I may have missed some important step.

**Comment by c-geek** [comment]
Indeed you missed an important point. Because of course I thought about key delegation, it is a near obvious solution when talking about keys' security.

My main issue with key delegation is: if I am able to delegate some key, why wouldn't I delegate my "vote" to Google keys, letting Google decide for me as it would be able to have a lot of keys delegated to its servers? (I take Google as an example of big structure able to provide a tremendous amount of CPU power).

With personal keys (UK), at least *you are not encouraged to share your key* since your identity and your UD could be stolen. The fact *your personal key* is required to compute a bloc combined to your repulsion of sharing your identity key is what lead us to a near "1 human = 1 vote" system.

Of course, someone may intentionally (or not) share its UK. But then, the rotation mechanism takes place and avoid this eventual nuisible node to take the control of the blockchain.

The ultimate goal of UK in nodes is just that: limit the potential nuisible actors. If you have a better idea, that would be welcomed.

**Comment by Sylvain303** [comment]
OK, I see. Thanks for mentioning this information. I will think a bit more about it and come back later, if I find a proposal. 

I will go deeper on PoW mentioned here: https://forum.ucoin.io/t/anatomy-of-bitcoins-blockchain/19/17
And a kind of "trust" on the delegation, may it could be enforced in some way.

I see your point about commitment of running a node with its own human credential, but I still disagree about hardcoding them on the node filesystem.

**Comment by c-geek** [comment]
If you disagree with the idea, you also disagree with how SSL keys are managed by webservers providing HTTPS. Or did I miss something and private keys are not stored clearly on webservers?

**Comment by M5oul** [comment]
[CLI arguments privacy](https://bitcoin.org/en/release/v0.13.0#bitcoin-cli-arguments-privacy) could be a solution and credentials could be stored on memory/RAM.

**Comment by M5oul** [comment]
May be [an interesting document](https://paragonie.com/blog/2016/02/how-safely-store-password-in-2016#nodejs).

**Comment by GammaNu** [comment]
A quick short term fix could be to remove salt and passwd because only pair.pub and pair.sec are needed.
We still have pair.sec stored in an improvable way, but no more passwd. This way with compromised server will have his private key compromised but not his original password (so if he use it elsewhere, only the private key need to be changed)

```
{
"salt": "monId",
 "passwd": "monMotDePassEnClair",
"pair": {
  "pub": "maClefPublic",
  "sec": "maClefPrivée"
 }
}
```

**Comment by Insoleet** [comment]
I don't think you can change the private key without changing the public key ?

**Comment by c-geek** [comment]
@Insoleet I think GammaNu only asks if we could remove salt/passwd from the conf.json file once the pub/sec keys have been derived from them.

**Comment by GammaNu** [comment]
Yep ;)

**Comment by c-geek** [comment]
With #739, we will be able to create modules providing the credentials by registering at the `config` phase. For example, we could have:

* a module adding the `--credentials` option as proposed in #435
* a module asking for the credentials at duniter startup in the terminal
* a module cleaning the configuration from salt/password and prefering to store the derived keys
* ...

I still have to define those phases, but at least this way we can choose the option we prefer.

**Comment by c-geek** [comment]
@GammaNu's proposal is now default: once the keypair is loaded, salt/password are cleaned from memory and configuration file. So we no more store them, but we also clean the existing conf file.

I still have to develop `--credentials` option. Or, maybe that will be a challenge for you guys :grin: 

Here is [a teaser](https://github.com/duniter/duniter-keypair/blob/master/index.js#L53-L69) on how to load keypair from duniter CLI options.

**Comment by Insoleet** [comment]
Next step : secure the private key in memory ;) http://stackoverflow.com/questions/1263350/cryptography-best-practices-for-keys-in-memory

**Comment by c-geek** [comment]
I have developed the `--keyfile` option (this name replaces `--credientials`). It will be available on Duniter@0.90.0.

So I close this particular issue, but for more extreme security improvements: please open another issue. :grin:

**Comment by c-geek** [comment]
For more informations about these options: https://github.com/duniter/duniter-keypair
